integration/integration_test.go

```diff
@@ -4854,6 +4830,86 @@ func (s *IntSuite) TestBPFSessionDifferentiation(c *check.C) {
 	c.Fatalf("Failed to find command events from two different sessions.")
 }
 
+// TestNodeRegistrationMethods tests registration of a node
+// with auth server and via proxy.
+func (s *IntSuite) TestNodeRegistrationMethods(c *check.C) {
+	tr := utils.NewTracer(utils.ThisFunction()).Start()
+	defer tr.Stop()
+
+	var testCases = []struct {
+		comment         string
+		getRegisterPort func(i *TeleInstance) string
+		logPattern      string
```

Comment:
matching against logs is fragile
instead, fetch the list of all nodes from the API (something like `t.GetSiteAPI().GetNodes()`) and check `node.GetUseTunnel()` to see if it connected over a tunnel (proxy) or auth.

Reply:
The motivation here is to cover the changes introduced in #5182, namely to assure that:
(a) users are not presented with unnecessary `ERROR` level messages as in #4663,
(b) the proxy mode is attempted as the first registration method in certain scenarios (which does not preclude falling back to auth) while "auth first, fall back to proxy" remains the default.

Merely verifying the end state would not achieve these goals. Checking the logs seemed to be the most straightforward way to do that.

Reply:
can you intercept connection attempts on the server or the client in any way?
like a wrapper around `net.Dialer` or `net.Listener`?

we're creating a bit of maintenance burden by checking logs with regexps, especially considering how slow and flaky our integration tests are.
next time someone changes the log message format, they will likely see integration tests fail (after ~15min), re-run them 1-2 more times (another ~30min) before looking at the errors and actually updating the test.